infer/environment.cc

```diff
@@ -599,14 +602,15 @@ void Environment::mergeWith(core::Context ctx, const Environment &other, core::L
             pair.second.knownTruthy = pair.second.knownTruthy && other.getKnownTruthy(var);
         } else {
             thisTO = otherTO;
+            sameType = false;
```

Comment:
I think this can be check if `otherTO` is niclass if we want a bit more perf, but I don't expect it to be noticeable.